### PR TITLE
Better management of locked funds

### PR DESCRIPTION
**Pull request by Krigpl**
Transaction system manages locked funds, not `FundsLocker`. `FundsLocker` is now just a simple class managing number of locked subtasks of a task.
This way everything funds related belongs in a single place - `TransactionSystem`.
No need to provide `locked` arguments in methods like `withdraw` or `concent_deposit`.
No need to remember about `FundsLocker` when querying for balance.

### CODE REVIEW

File: golem/transactions/ethereum/ethereumtransactionsystem.py [hunk 1]
```diff
@@ -102,18 +105,72 @@ def get_payment_address(self):
...
+                gnt / denoms.ether,
+                self._gntb_locked / denoms.ether,
+            ))
+        if num > self._payments_locked:
+            raise Exception("Can't unlock {} payments, locked: {}".format(
+                num,
+                self._payments_locked,
+
+            ))
+        log.info("Unlocking %f GNT and ETH for %d payments", gnt, num)
```
Comment (maaktweluit) [inline golem/transactions/ethereum/ethereumtransactionsystem.py]: Where is the ETH unlocked in this function?

File: golem/transactions/ethereum/ethereumtransactionsystem.py [hunk 2]
```diff
@@ -102,18 +105,72 @@ def get_payment_address(self):
...
+
+        eth = self._eth_for_batch_payment(num)
+        if self._payments_locked == 0 and \
+           self.payment_processor.reserved_eth == 0:
+            eth += self._eth_base_for_batch_payment()
+        eth_available = self.get_available_eth()
+        if eth > eth_available:
+            raise NotEnoughFunds(eth, eth_available, 'ETH')
+
+        log.info("Locking %f GNT and ETH for %d payments", gnt, num)
```
Comment (maaktweluit) [inline golem/transactions/ethereum/ethereumtransactionsystem.py]: Where is the ETH locked in this function?

File: tests/golem/transactions/ethereum/test_ethereumtransactionsystem.py [hunk 3]
```diff
@@ -25,9 +25,11 @@ class TestEthereumTransactionSystem(TestWithDatabase, LogTestCase,
     def setUp(self):
         super().setUp()
         self.sci = Mock()
-        self.sci.GAS_PRICE = 0
-        self.sci.GAS_BATCH_PAYMENT_BASE = 0
+        self.sci.GAS_PRICE = 10 ** 9
```
Comment (maaktweluit) [inline tests/golem/transactions/ethereum/test_ethereumtransactionsystem.py:29]: Repeated magic numbers, maybe they can have a common source?

File: tests/golem/transactions/ethereum/test_ethereumtransactionsystem.py [hunk 4]
```diff
@@ -179,9 +186,44 @@ def test_withdraw(self):
 
         # Not enough GNT with lock
         with self.assertRaises(NotEnoughFunds):
-            self.ets.withdraw(gnt_balance + gntb_balance - 1, dest, 'GNT', 2)
+            self.ets.withdraw(gntb_balance + locked_gnt + 1, dest, 'GNT')
         self.sci.reset_mock()
 
+    def test_locking_funds(self):
+        eth_balance = 10 * 10 ** 18
```
Comment (maaktweluit) [inline tests/golem/transactions/ethereum/test_ethereumtransactionsystem.py]: same here, isnt it better to use denoms.ether?

**Review by maaktweluit — COMMENTED** [review]
LGTM!

Some q's, will approve on reply or fix :)

File: golem/transactions/ethereum/ethereumtransactionsystem.py (hunk 2 above)
Comment (Krigpl) [reply golem/transactions/ethereum/ethereumtransactionsystem.py]: Locked ETH is calculated based on the `_payments_locked` variable.

File: golem/transactions/ethereum/ethereumtransactionsystem.py (hunk 1 above)
Comment (Krigpl) [reply golem/transactions/ethereum/ethereumtransactionsystem.py]: Ditto

File: tests/golem/transactions/ethereum/test_ethereumtransactionsystem.py (hunk 3 above)
Comment (Krigpl) [reply tests/golem/transactions/ethereum/test_ethereumtransactionsystem.py:29]: The value doesn't really matter and is not used anywhere else (as in changing this value doesn't trigger changing any other `10**9`), that's why imo it's fine like this.

File: tests/golem/transactions/ethereum/test_ethereumtransactionsystem.py (hunk 4 above)
Comment (Krigpl) [reply tests/golem/transactions/ethereum/test_ethereumtransactionsystem.py]: This is somewhat arbitrary as well but will use `denoms.ether` for clarity

File: golem/transactions/ethereum/ethereumtransactionsystem.py (hunk 2 above)
Comment (maaktweluit) [reply golem/transactions/ethereum/ethereumtransactionsystem.py]: ahhh, now i see, its all there :D

File: golem/transactions/ethereum/ethereumtransactionsystem.py (hunk 1 above)
Comment (maaktweluit) [reply golem/transactions/ethereum/ethereumtransactionsystem.py]: thanks! ditto

**Review by maaktweluit — APPROVED** [review]

File: golem/transactions/ethereum/ethereumtransactionsystem.py [hunk 5]
```diff
@@ -102,18 +105,80 @@ def get_payment_address(self):
...
-        gnt_total = self._gnt_balance + self._gntb_balance
-        gnt_av = gnt_total - self.payment_processor.reserved_gntb
-        return gnt_total, gnt_av, self._eth_balance, \
-            self._last_gnt_update, self._last_eth_update
+    def get_available_eth(self) -> int:
+        return max(self._eth_balance - self.get_locked_eth(), 0)
+
+    def get_locked_eth(self) -> int:
+        eth = self.payment_processor.reserved_eth + \
+            self._eth_for_batch_payment(self._payments_locked)
```
Comment (mfranciszkiewicz) [inline golem/transactions/ethereum/ethereumtransactionsystem.py]: Is eth for batch payment always a part of locked eth? What about providers?

Comment (Krigpl) [reply golem/transactions/ethereum/ethereumtransactionsystem.py]: ETH for batch payments has two parts: base and per payment. This only adds the per payment part which for providers will be 0. The logic below checks whether the base should be added.

**Pull request merged**